torch/csrc/jit/passes/memory_planning.cpp

```diff
@@ -0,0 +1,399 @@
+  auto ttp = value.type()->cast<TensorType>();
+  if (!ttp) {
+    TORCH_WARN(
+        node_header,
+        "; ",
+        value.debugName(),
+        " isn't a tensortype: ",
+        *value.type());
+    return c10::nullopt;
+  }
+  if (!ttp->scalarType().has_value()) {
+    TORCH_WARN(
+        node_header,
+        "; ",
+        value.debugName(),
+        " was profiled but didn't have a scalar type: ",
+        *ttp);
+    return c10::nullopt;
+  }
+  if (!ttp->sizes().concrete_sizes().has_value()) {
+    TORCH_WARN(
+        node_header,
+        "; ",
+        value.debugName(),
+        " was profiled but doesn't have sizes: ",
+        *ttp);
+    return c10::nullopt;
+  }
+
+  auto scalar_type = ttp->scalarType();
+  if (!scalar_type.has_value()) {
+    TORCH_WARN(
+        node_header,
+        "; ",
+        value.debugName(),
+        " doesn't have a scalar type: ",
+        *ttp);
+    return c10::nullopt;
+  }
+
+  auto element_size = c10::elementSize(scalar_type.value());
+  // TODO: when does this fail? answer: in place mutation
+  auto numel = ttp->numel();
+  if (!numel.has_value()) {
+    TORCH_WARN(
+        node_header, "; ", value.debugName(), " doesn't have numel: ", *ttp);
+    return c10::nullopt;
+  }
+
+  return numel.value() * element_size;
+}
+
+Node* insertSlabNode(
+    const std::shared_ptr<Graph>& graph,
+    size_t total_size,
+    c10::optional<at::Device> device_type = c10::nullopt) {
+  auto slab = graph->create(prim::AllocateSlab, 1);
+  slab->output()->setType(StorageType::get());
+  slab->i_(attr::total_size, (int64_t)total_size);
+
+  if (device_type || (device_type = jit::tensorexpr::pickDeviceType(graph))) {
+    slab->i_(attr::device_type, static_cast<int8_t>(device_type->type()));
+  } else {
+    slab->i_(attr::device_type, static_cast<int8_t>(at::kCPU));
+  }
+  slab->insertBefore(graph->nodes().front());
+  return slab;
+}
+
+void insertAllocNodes(
+    const std::shared_ptr<Graph>& graph,
+    size_t total_size,
+    std::vector<std::pair<const Value*, MemRegion>>& managed_value_allocations,
+    c10::optional<at::Device> device_type = c10::nullopt) {
+  auto slab = insertSlabNode(graph, total_size, device_type);
+  auto release_slab = graph->create(prim::ReleaseSlab, 0);
+  release_slab->addInput(slab->output());
+
+  for (auto& item : managed_value_allocations) {
+    auto val = item.first;
+    auto region = item.second;
+
+    // const_cast fishy?
+    Node* node = const_cast<Node*>(val->node());
+
+    // the way that this node magically *becomes* the out varaint is simply
+    // by add an extra input. this is because op resolution happens
+    // at runtime via the op registry (by matching on the schema).
+    auto* alloc = graph->create(prim::AllocateTensor, 1);
+    node->addInput(alloc->output());
+    alloc->insertBefore(node);
+    alloc->addInput(slab->output());
+
+    auto ttp = val->type()->expect<c10::TensorType>();
+    alloc->ty_(attr::profiled_type, ttp);
```

Comment:
Why do we need the type here as an attribute of allow ?

Reply:
`AllocTensor` uses the type info to set the shapes/strides/etc

```
                 *type->sizes().concrete_sizes(),
                 *type->strides().concrete_sizes(),
                 at::TensorOptions(*type->device()).dtype(*type->scalarType()));
```

whether that's necessary i'm not sure, but it is being used

Reply:
as a future reference I would consider storing the sizes/strides & dtype/device/TensorOptions directly, instead of requiring the extra indirection since `AllocTensor` might be called many times in a hot loop

Reply:
@eellison I had that before and replaced it with this because I was basically unpacking, repacking, unpacking again `TensorTypePtr`. I can put it back if you think it's a real performance issue but conceptually it seems more straightforward to leave as is (especially since the attributes can't get out of sync this way)

Reply:
fine to leave out of this PR, let's just make sure to catalogue good potential follow ups